pallets/vrf/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The relay block hash type (probably H256)
+		type RelayBlockHash: Parameter
+			+ Member
+			+ MaybeSerializeDeserialize
+			+ Default
+			+ Copy
+			+ AsRef<[u8]>;
+		/// Gets the most recent relay block hash and relay slot number in `on_initialize`
+		type MostRecentVrfInputGetter: GetMostRecentVrfInputs<Self::RelayBlockHash, Slot>;
+		/// Convert account to VRF key, presumably via AuthorMapping instance
+		/// TODO: maybe TryConvert and map to error because fallible conversion
+		type AccountToVrfId: Convert<Self::AccountId, AuthorityId>;
+		/// Whether or not the round changed this block
+		type RoundChanged: RoundChangedThisBlock;
+		/// Get the selected candidate accounts from staking
+		type SelectedCandidates: Get<Vec<Self::AccountId>>;
+	}
+
+	/// Current block randomness
+	/// Set in `on_initialize`, before it will contain the randomness from the last block
+	#[pallet::storage]
+	#[pallet::getter(fn current_randomness)]
+	pub type CurrentRandomness<T> = StorageValue<_, MaybeRandomness, ValueQuery>;
+
+	/// Last block randomness
+	/// Set in `on_initialize`, before it will contain the randomness from the last last block
+	#[pallet::storage]
+	#[pallet::getter(fn last_randomness)]
+	pub type LastRandomness<T> = StorageValue<_, MaybeRandomness, ValueQuery>;
+
+	/// Current set of authorities by AuthorityId
+	/// Set in `on_initialize` upon round changes
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub type Authorities<T> = StorageValue<_, Vec<AuthorityId>, ValueQuery>;
+
+	/// Most recent relay chain block hash
+	/// Set in `on_initialize` before setting randomness
+	#[pallet::storage]
+	#[pallet::getter(fn most_recent_relay_block_hash)]
+	pub(crate) type MostRecentRelayBlockHash<T: Config> =
+		StorageValue<_, T::RelayBlockHash, ValueQuery>;
+
+	/// Most recent relay chain slot number
+	/// Set in `on_initialize` before setting randomness
+	#[pallet::storage]
+	#[pallet::getter(fn most_recent_relay_slot_number)]
+	pub(crate) type MostRecentSlotNumber<T> = StorageValue<_, Slot, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+			let initial_weight = Self::set_authorities_if_round_changed();
+			let (used_weight, relay_block_hash, relay_slot_number) =
+				Self::set_most_recent_vrf_inputs(initial_weight);
+			Self::set_randomness(used_weight, relay_block_hash, relay_slot_number)
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Set authorities if the round has changed
+		fn set_authorities_if_round_changed() -> Weight {
+			if T::RoundChanged::round_changed_this_block() {
+				let selected_candidates = T::SelectedCandidates::get();
+				let returned_weight: Weight = selected_candidates.len() as u64
+					* T::DbWeight::get().read
```

Comment:
every selected candidate requires a read for `T::AccountToVrfId::convert(x)` (below), we will need that conversion in storage or somewhere

Reply:
Oh I see, then my suggestion before still applies. In the pallet we are assuming that this conversion involves a storage read, but that might not be the case for other runtimes (maybe because Account & VRFId are the same)? We could define a custom trait instead of Convert that also returns the weight. Maybe this is overkilll, but I am just trying to think of other potential users of the pallet.

I am also fine leaving this as it is. If we do so, we need to document very well where the reads come from

Reply:
Also in one branch the db read for the round change seems not to be counted